src/lib/statistics/statistics_objects/top_k_as_generic_histogram.cpp

```diff
@@ -0,0 +1,129 @@
+#include "top_k_as_generic_histogram.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "equal_distinct_count_histogram.hpp"
+#include "generic_histogram_builder.hpp"
+
+namespace opossum {
+
+template <typename T>
+std::shared_ptr<GenericHistogram<T>> TopKAsGenericHistogram<T>::from_column(const Table& table,
+                                                                            const ColumnID column_id,
+                                                                            const HistogramDomain<T>& domain) {
+  auto value_distribution = value_distribution_from_column(table, column_id, domain);
+
+  if (value_distribution.empty()) {
+    return nullptr;
+  }
+
+  // If the column holds less than K distinct values use the distinct count as TOP_K instead.
+  const auto k = std::min(TOP_K_DEFAULT, value_distribution.size());
+
+  std::vector<T> top_k_names(k);
+  std::vector<HistogramCountType> top_k_counts(k);
+
+  // Bring Top K values (with highest occurrences counts) to the front (first k elements) of the value distribution.
+  auto top_k_front_value_distribution = value_distribution;
+  std::nth_element(top_k_front_value_distribution.begin(), top_k_front_value_distribution.begin() + k,
+                   top_k_front_value_distribution.end(),
+                   [&](const auto& left_value_count, const auto& right_value_count) {
+                     return left_value_count.second > right_value_count.second;
+                   });
+
+  // Sort Top K values lexicographically
+  // We later use the lexicographically sorted Top K values for easier and more performant histogram construction.
+  std::sort(top_k_front_value_distribution.begin(), top_k_front_value_distribution.begin() + k,
+            [&](const auto& left_value_count, const auto& right_value_count) {
+              return left_value_count.first < right_value_count.first;
+            });
+
+  for (auto top_k_index = 0u; top_k_index < k; top_k_index++) {
+    top_k_names[top_k_index] = top_k_front_value_distribution[top_k_index].first;
+    top_k_counts[top_k_index] = top_k_front_value_distribution[top_k_index].second;
+  }
+
+  // Remove Top K values from value distribution
+  for (auto top_k_index = 0u; top_k_index < k; top_k_index++) {
+    auto value_distribution_it = remove(value_distribution.begin(), value_distribution.end(),
+                                        std::make_pair(top_k_names[top_k_index], top_k_counts[top_k_index]));
+    value_distribution.erase(value_distribution_it, value_distribution.end());
+  }
+
+  // Each Top K value is modeled as one bin with height as its stored count.
```

Comment:
"Each Top K value is represented by a single bin whose height refers to the value's occurrences."?